servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java

```diff
@@ -227,75 +250,176 @@ private boolean shouldRedirect(final int redirectCount, final StreamingHttpRespo
+            return redirectSingle.config.changePostToGet &&
+                    (statusCode == MOVED_PERMANENTLY.code() || statusCode == FOUND.code()) &&
+                    POST.name().equals(originalMethod.name()) ? GET : originalMethod;
+        }
+
+        private static boolean isRelative(final HttpRequestMetaData originalRequest,
+                                          @Nullable final String originalScheme,
+                                          final HttpRequestMetaData redirectRequest) {
+            final String toHost = redirectRequest.host();
+            if (toHost == null) {
+                return true;
+            }
+            final HostAndPort original = originalRequest.effectiveHostAndPort();
+            if (original == null) {
+                return false;   // Can not extract host and port from the original request => no guarantee it's relative
+            }
+            if (!toHost.equalsIgnoreCase(original.hostName())) {
+                return false;
+            }
+            return inferPort(redirectRequest.port(), redirectRequest.scheme(), original.port()) ==
+                    inferPort(original.port(), originalScheme, original.port());
+        }
+
+        private static int inferPort(final int parsedPort, @Nullable final String scheme,
+                                     final int fallbackPort) {
+            if (parsedPort >= 0) {
+                return parsedPort;
+            }
+            if (scheme == null) {
+                return fallbackPort;
+            }
+            return "https".equalsIgnoreCase(scheme) ? 443 : 80;
+        }
+
+        private static void copyAll(final StreamingHttpRequest originalRequest,
+                                    final StreamingHttpRequest redirectRequest) {
+            redirectRequest.setHeaders(originalRequest.headers());
+            redirectRequest.transformMessageBody(p -> p.ignoreElements().concat(originalRequest.messageBody()));
+            // Use `transform` to update PayloadInfo flags, assuming trailers may be included in the message body
+            redirectRequest.transform(NOOP_TRAILERS_TRANSFORMER);
+            // FIXME: instead of `transform`, preserve original PayloadInfo/FlushStrategy when it's API is available
+        }
+
+        private void copySome(final StreamingHttpRequest request, final StreamingHttpRequest redirectRequest) {
+            // NOTE: for security reasons we do not copy any headers or payload body from the original request by
+            // default for non-relative redirects.
+            for (CharSequence headerName : redirectSingle.config.headersToRedirect) {
+                for (CharSequence headerValue : request.headers().values(headerName)) {
+                    redirectRequest.addHeader(headerName, headerValue);
+                }
+            }
+
+            final Config config = redirectSingle.config;
+            if (config.redirectPayloadBody) {
```

Comment:
should this be conditional based upon the request/response?

Reply:
If users need more flexibility and control, they can use `prepareRequest` that gives access to the original request, redirect response, and redirect request. This API is intended to be simple for cases when they know what has to be copied.

Reply:
this is now in `RedirectRequestTransformer` so users have control 💪 🏅